velox/exec/VectorHasher.cpp

```diff
@@ -247,36 +246,43 @@ bool VectorHasher::makeValueIdsDecoded(
   auto values = decoded_.data<T>();
 
   bool success = true;
-  rows.applyToSelected([&](vector_size_t row) INLINE_LAMBDA {
+  int decodedBaseVisited = 0;
+  rows.testSelected([&](vector_size_t row) {
     if constexpr (mayHaveNulls) {
       if (decoded_.isNullAt(row)) {
         if (multiplier_ == 1) {
           result[row] = 0;
         }
-        return;
+        return true;
       }
     }
+
     auto baseIndex = indices[row];
-    uint64_t id = cachedHashes_[baseIndex];
-    if (id == 0) {
-      T value = values[baseIndex];
-
-      if (!success) {
-        // If all were not mappable we just analyze the remaining so we can
-        // decide the hash mode.
-        analyzeValue(value);
-        return;
+    uint64_t& id = cachedHashes_[baseIndex];
+
+    if (success) {
+      if (id == 0) {
+        T value = values[baseIndex];
+        id = valueId(value);
+        decodedBaseVisited++;
+        if (id == kUnmappable) {
+          analyzeValue(value);
+          success = false;
+        }
       }
-      id = valueId(value);
-      if (id == kUnmappable) {
-        analyzeValue(value);
-        success = false;
-        return;
+      result[row] = multiplier_ == 1 ? id : result[row] + multiplier_ * id;
+    } else {
+      if (id == 0) {
+        id = kUnmappable;
+        decodedBaseVisited++;
+        analyzeValue(values[baseIndex]);
       }
-      cachedHashes_[baseIndex] = id;
     }
-    result[row] = multiplier_ == 1 ? id : result[row] + multiplier_ * id;
+
+    bool shouldContinue = success || decodedBaseVisited < cachedHashes_.size();
```

Comment:
My understanding of the problem is that analyzeValue is called repeatedly on all remaining rows as we do not track rows we already analyzed using cachedHashes_.

If that's the case, then why not start tracking analyzed rows in cachedHashes_ by setting the hash to kUnmappable.

Using a simple counter (decodedBaseVisited) doesn't work in cases when DictionaryVector refers to a subset of indices in the base vector. In this case we'll never reach cachedHashes_.size() and the optimization won't apply.

Reply:
@mbasmanova 
Let me know if I'm understanding correctly your suggestion; but it does sound similar to a scenario that I already considered when I implemented this optimization.

> Using a simple counter (decodedBaseVisited) doesn't work in cases when DictionaryVector refers to a subset of indices in the base vector. In this case we'll never reach cachedHashes_.size() and the optimization won't apply.

In the case where not all base values are reference, we will have "no choice" but to scan the entire (outer) input because without doing so, we cannot be sure whether all base values are referenced or not.

Reply:
But we do not need to call analyzeValues for each value, no?

Reply:
Consider this simple case:

Inner dictionary has 3 distinct values.

Input vector (length 1000): first 999 dict index refers to 1st distinct value; 1000th dict index refers to 2nd distinct value. The 3rd distinct value is never referenced.

VectorHasher immediately fails to "compute value" and enters "failure/analyze" mode. In this situation, no optimization is possible. We will have to scan the entire 1000 input by virtue of the fact that the 3rd distinct value is never referenced.

I thought of this problem when making this optimization. Upon realizing this problem, one might be tempted to say: why not just always analyze the inner dict vector and "map the values back" to the outer input vector. Such a solution would not be work because it's possible that the inner dictionary vector is much larger (say 1000 values) whilst the outer value is tiny (say, length of 5). In this case, this strategy would actually result in a regression.

Thus, after much thinking, I felt the current optimization, as is, is about as much as we can get away with, without sacrificing other cases.

Reply:
I see. My assumption is that CPU times goes not into looping over 1000 rows, but into calling analyzeValue 1000 times. Is this not the case? If we loop over 1000 rows, but only call analyzeValue twice, will this not be more efficient than current code?

Reply:
> But we do not need to call analyzeValues for each value, no?

Yes, analyzeValue is the main CPU consumer, due to putting things into F14set.  Perhaps can you illustrate your suggestion with an example so I may more concretely understand and comment on its merits?

Reply:
I'm wrapping up my day and will be out till next Tue. Please, ping me on Tue to continue.

Reply:
@mbasmanova hello masha, I was ooo yesterday. Just pinging to continue this discussion. Thank you!